### PR TITLE
ci: Create composite action for maximizing the build space

### PR DESCRIPTION
**Pull request by LKreutzer**
Signed-off-by: Lars Kreutzer <lars.kreutzer@tngtech.com>

## Summary

- A composite GH action is created to maximize the build space in the GH runners.

## Test Plan

- Compare workflow logging on PR to logging on master.
  - Build space left at the end of the workflows is more than before the action cleaned up. 

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: .github/workflows/composite/maximize-build-space/action.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+---
+name: maximize-build-space
+description: Maximize build space on github runner
```
Comment (Neudrino) [inline .github/workflows/composite/maximize-build-space/action.yml]: Should the name `Github` be capitalized?

File: .github/workflows/composite/maximize-build-space/action.yml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+runs:
+  using: composite
+  steps:
+    - name: Maximize build space
+      shell: bash
+      run: |
+        echo "Available storage before:"
+        df -h
+        echo "Removing /usr/share/dotnet, /usr/local/lib/android, /opt/ghc"
```
Comment (Neudrino) [inline .github/workflows/composite/maximize-build-space/action.yml:15]: Is that not redundant documentation?

File: .github/workflows/composite/maximize-build-space/action.yml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+        sudo rm -rf /usr/share/dotnet
+        sudo rm -rf /usr/local/lib/android
+        sudo rm -rf /opt/ghc
```
Comment (Neudrino) [inline .github/workflows/composite/maximize-build-space/action.yml]: - Do we have no password sudo permissions on the runner?
- Should we delete all folders in one call using line breaks?

**Review by jheidbrink — APPROVED** [review]

File: .github/workflows/agw-workflow.yml [hunk 4]
```diff
@@ -146,20 +146,10 @@ jobs:
...
-          echo "Removing /usr/share/dotnet, /usr/local/lib/android, /opt/ghc"
-          sudo rm -rf /usr/share/dotnet
-          sudo rm -rf /usr/local/lib/android
-          sudo rm -rf /opt/ghc
-          echo "Available storage after:"
-          df -h
       - name: Check Out Repo
         # This is necessary for overlays into the Docker container below.
         uses: actions/checkout@7884fcad6b5d53d10323aee724dc68d8b9096a2e # pin@v2
+      - uses: ./.github/workflows/composite/maximize-build-space
```
Comment (jheidbrink) [inline .github/workflows/agw-workflow.yml]: Nitpick: All the other steps here have a `name: ` entry with a descriptive name, we could adhere to that pattern.

File: .github/workflows/composite/maximize-build-space/action.yml (hunk 2 above)
Comment (LKreutzer) [reply .github/workflows/composite/maximize-build-space/action.yml:15]: The `rm` calls take ~1min which does not produce any logging, so I think it is justified to print one line to give feedback.

File: .github/workflows/composite/maximize-build-space/action.yml [hunk 5]
```diff
@@ -0,0 +1,18 @@
+---
+name: maximize-build-space
```
Comment (jheidbrink) [inline .github/workflows/composite/maximize-build-space/action.yml:2]: Nitpick: We could include a short rationale why we don't use an existing action for this (as far as I'm concerned the reason is that this one does less and is simpler)

File: .github/workflows/composite/maximize-build-space/action.yml (hunk 1 above)
Comment (LKreutzer) [reply .github/workflows/composite/maximize-build-space/action.yml]: Done

File: .github/workflows/composite/maximize-build-space/action.yml (hunk 3 above)
Comment (LKreutzer) [reply .github/workflows/composite/maximize-build-space/action.yml]: 1. `sudo` is needed
2. Done

File: .github/workflows/agw-workflow.yml (hunk 4 above)
Comment (LKreutzer) [reply .github/workflows/agw-workflow.yml]: Done. (Currently almost all local action calls do not have a name field.)

File: .github/workflows/composite/maximize-build-space/action.yml (hunk 5 above)
Comment (LKreutzer) [reply .github/workflows/composite/maximize-build-space/action.yml:2]: Done

File: .github/workflows/composite/maximize-build-space/action.yml (hunk 3 above)
Comment (Neudrino) [reply .github/workflows/composite/maximize-build-space/action.yml]: Sorry I missed a dash in the question. It should have been:
1. Do we have no-password sudo permissions on the runner?
I.e. if we can use `sudo` for the active user without password confirmation.

Comment (LKreutzer) [reply .github/workflows/composite/maximize-build-space/action.yml]: Yes, it works as it is.

**Review by Neudrino — APPROVED** [review]

**Pull request merged**